draft-kazuho-httpbis-priority.md

```diff
@@ -98,6 +98,70 @@ The terms sh-token and sh-boolean are imported from
+Section 5.3, which some implementers have opted not to fully support.
+The lack of signalling about the status of the implementation has
+caused several implementations to implement heuristics to detect when
+the clients they are connected to do not support priorities as
+defined and take steps to compensate for that.
+
+The intent of this negotiation is to provide a signalling mechanism
+for each peer to communicate which, if any, priority schemes are
+supported, as well as the server's ranked preference.
+
+For both HTTP/2 and HTTP/3, either peer's SETTINGS may arrive first,
+so any negotiation must be unilateral and not rely upon receiving
+the peer's SETTINGS value.
+
+Implementations are likely to only use one prioritization scheme at
+once, and may be unable to change the scheme once established, so the
+setting MUST be sent prior to the first request if it is ever sent.
+In HTTP/3, SETTINGS may arrive after the first request even if
+they are sent first.  In order to avoid the server incorrectly
+choosing a priority scheme in HTTP/3, the client SHOULD only send
+priority information prior to receiving the server's SETTINGS if
+it only offers a single scheme.  If it offers multiple schemes,
+the client SHOULD delay sending priority information until
+after it knows what the server supports, or risk the server
+choosing a sub-optimal scheme.
+
+## The SETTINGS_PRIORITIES SETTINGS Parameter
+
+This document adds a new SETTINGS parameter to those defined by
+[RFC7540], Section 6.5.2.
+
+The new parameter name is SETTINGS_PRIORITIES, which allows both
+peers to indicate which prioritization schemes they support.
+
+A value of 0 indicates no support for priorities. If either side sends the
+parameter with a value of 0, clients SHOULD NOT send priority frames
+and servers SHOULD NOT make any assumptions based on the presence or
+lack thereof of priority frames.
+
+If the value is non-zero, then the least significant 8 bits indicates the
+peer's preferred priority scheme, the second least significant 8 bits
+indicates the peer's second choice, and so on.  This allows expressing
+support for 4 schemes in HTTP/2 and 7 in HTTP/3.  If any octet is 0,
+all more significant octets MUST also be 0.
+
+In HTTP/2, the setting SHOULD appear in the first SETTINGS frame and peers
+MUST NOT process the setting if it's received multiple times in order to
+avoid changing the agreed upon prioritization scheme.
+
+If there is a prioritization scheme supported by both the client and server,
+then the servers's preference order prevails and both peers SHOULD
+only use the agreed upon priority scheme for the remainder of the session.
+The server chooses because it is in the best position to know what
+information from the client is of the most value.
+
+An 8 bit value of 1 in HTTP/2 indicates support for HTTP/2 priorities
+as defined in Section 5.3 of [RFC7540] and is an error in HTTP/3 because
+there is not a clean mapping to HTTP/3.
```

Comment:
I think I'd like to see values defined a little more formally. For example, a proper noun for these things (priorty scheme identifier?), a human-friendly identifer name (that we can bikeshed on) and a codepoint. Then I would restate things as:

"This document defines the priority scheme identifier H2_TREE (8-bit value of 1) that indicates support for HTTP/2-style priorities. 

The H2_TREE priority scheme identifier MUST NOT be be sent in an HTTP/3 settings because there is no defined mapping of this scheme. Endpoints MUST treat receipt of H2_TREE as a connection error of type H3_SETTINGS_ERROR."

I also think there is value in an IANA registry for these identifiers.